### PR TITLE
Refine cells taking into account width, length, height

### PR DESCRIPTION
**Pull request by aritorto**
A new method has been added in Geometry, to refine a single cell taking into account different widths, lengths, and heights.  It generalizes the previous refine() whose refinement is uniform.

### CODE REVIEW

**Comment by aritorto** [comment]
jenkins build this please

File: opm/grid/cpgrid/Geometry.hpp [hunk 1]
```diff
@@ -986,6 +986,425 @@ namespace Dune
...
+                for (int j = 0; j < refined_dim[1] +1; ++j) {
+                    for (int i = 0; i < refined_dim[0] +1; ++i) {
+                        for (int k = 0; k < refined_dim[2] +1; ++k) {
+                            // Compute the index of each global refined corner associated with 'jik'.
+                            int refined_corner_idx =
+                                (j*(refined_dim[0]+1)*(refined_dim[2]+1)) + (i*(refined_dim[2]+1)) + k;
+                            // Compute the local refined corner of the unit/reference cube associated with 'jik'.
+                            const double local_x = localCoordNumerator(dx, i/cells_per_dim[0], double((i % cells_per_dim[0])) / cells_per_dim[0]);
+                            const double local_y = localCoordNumerator(dy, j/cells_per_dim[1], double((j % cells_per_dim[1])) / cells_per_dim[1]);
+                            const double local_z = localCoordNumerator(dz, k/cells_per_dim[2], double((k % cells_per_dim[2])) /cells_per_dim[2]);
```
Comment (blattms) [inline opm/grid/cpgrid/Geometry.hpp:1072]: This looks really complicated. It took me a long time to understand it. That might be me, of course. I would probably have done it like this (just keeping track of the coordinates by adding to them):
```suggestion
                double local_y = 0;
                for (int j = 0; j < refined_dim[1] +1; ++j, local_y+=dy[j]) {
                    double local_x = 0;
                    for (int i = 0; i < refined_dim[0] +1; ++i, local_x+=dx[i]) {
                        double local_z = 0;
                        for (int k = 0; k < refined_dim[2] +1; ++k, local_z+=dz[k]) {
                            // Compute the index of each global refined corner associated with 'jik'.
                            int refined_corner_idx =
                                (j*(refined_dim[0]+1)*(refined_dim[2]+1)) + (i*(refined_dim[2]+1)) + k;
                            // local_x, local_y, and local z should already be correct now.
```

**Review by blattms — CHANGES_REQUESTED** [review]
Just some comments. Printing should be removed before merging.

File: opm/grid/cpgrid/CpGridData.hpp [hunk 2]
```diff
@@ -318,6 +318,9 @@ class CpGridData
     /// @brief Check that every cell to be refined has cuboid shape.
     void checkCuboidShape(const std::vector<int>& cellIdx_vec) const;
 
+    /// @brief For selected cell indices, computes the variation in x-,y-, and z-direction, assuming each cell has cubiod shape.
+    std::array<std::vector<double>,3> getDxDyDz(const std::array<int,3>& startIJK, const std::array<int,3>& endIJK) const;
+
```
[the remark this answers is not in the document]
Comment (aritorto) [reply opm/grid/cpgrid/CpGridData.hpp:323]: Yes, that makes sense. Thanks!

File: opm/grid/cpgrid/Geometry.hpp [hunk 3]
```diff
@@ -986,6 +986,425 @@ namespace Dune
...
+                            std::cout << "local_x, y, z: " << local_x << " " <<  local_y << " "<< local_z << std::endl;
+                            const LocalCoordinate& local_refined_corner = { local_x/sum_dx, local_y/sum_dy, local_z/sum_dz };
+                            assert(local_x/sum_dx <= 1.);
```
[the remark this answers is not in the document]
Comment (aritorto) [reply opm/grid/cpgrid/Geometry.hpp:1075]: Shall I remove these asserts? Or add certain tolerance? For example, 1. + e-6 ?

File: opm/grid/cpgrid/CpGridData.cpp [hunk 4]
```diff
@@ -1808,6 +1808,67 @@ void CpGridData::checkCuboidShape(const std::vector<int>& cellIdx_vec) const
...
+        // x = |corn[1]-corn[0]|
+        std::vector<cpgrid::Geometry<0,3>::GlobalCoordinate> aFewCorners;
+        aFewCorners.resize(2); // {'0', '1'}
+        aFewCorners[0] = (*(this -> geometry_.geomVector(std::integral_constant<int,3>()))).get(cellToPoint[0]).center();
+        aFewCorners[1] = (*(this -> geometry_.geomVector(std::integral_constant<int,3>()))).get(cellToPoint[1]).center();
+
+        double  x;
+        x = std::sqrt( ((aFewCorners[1][0] -aFewCorners[0][0])*(aFewCorners[1][0] -aFewCorners[0][0])) +
+                       ((aFewCorners[1][1] -aFewCorners[0][1])*(aFewCorners[1][1] -aFewCorners[0][1])) +
+                       ((aFewCorners[1][2] -aFewCorners[0][2])*(aFewCorners[1][2] -aFewCorners[0][2])));
```
[the remark this answers is not in the document]
Comment (aritorto) [reply opm/grid/cpgrid/CpGridData.cpp]: Thanks! it's changed now

**Comment by aritorto** [comment]
jenkins build this please

**Pull request merged**